### PR TITLE
ensuring that 0 in mean_sd return NA

### PR DESCRIPTION
**Pull request by edwardhuh**
**Enhancements:**
- Added a test that ensures that counts with 0 in MultiIndex will return 'NA' for both mean and sd, even when `suppress=True`. 
- Preceded `.fillna('*')` with `.astype(object)` to ensure that  can be applied to more diverse input dataframes. 
- Created a randomly generated simple dataframe that is smaller than the `mega-store.csv` that should be more intuitive for readers to follow. Thinking of developing this into future documentation when approved.

### CODE REVIEW

File: src/zentables/__init__.py [hunk 1]
```diff
@@ -13,6 +13,6 @@
     df.zen.pretty()
 """
 
-from .zentables import ZenTablesAccessor, set_options
+from .zentables import ZenTablesAccessor, set_options, _do_suppression
```
Comment (khwilson) [inline src/zentables/__init__.py]: I thought we decided _not_ to make this part of the public API?

File: src/zentables/zentables.py [hunk 2]
```diff
@@ -519,7 +519,7 @@ def _internal_freq_table(
             mask = _do_suppression(pivot, low, high)
 
         if props is None and suppress:
-            return pivot.where(~mask).fillna("*")
+            return pivot.where(~mask).astype(object).fillna("*")
```
Comment (khwilson) [inline src/zentables/zentables.py]: Casting to type `object` seems like it's probably wrong?

File: src/zentables/zentables.py [hunk 3]
```diff
@@ -735,6 +735,8 @@ def mean_sd_table(
         count = pivot.xs("count", axis=1).astype("Int64").fillna(0)
         mean = pivot.xs("mean", axis=1)
         std = pivot.xs("std", axis=1)
+        # issues can arise when the shape is inconsistent because some groupings are entirely empty
+        assert count.shape == mean.shape == std.shape, "Ensure that all categories have at least some entry"
```
Comment (khwilson) [inline src/zentables/zentables.py]: You should raise a `ValueError` or somesuch. In general, `AssertionError`s are not meant to be caught, but you may wish to catch this one.

File: tests/test_suppression.py [hunk 4]
```diff
@@ -132,3 +133,26 @@ def test_single_row():
...
     mask = zen._do_suppression(df, low=1, high=5)
     assert (mask.values == expected_array).all()
+
+def test_nan_in_mean_sd_table():
+    total_length = 20
+    cuisine = ['Chinese', 'Korean', 'Italian']
+    city = ['Boston', 'Providence']
+    who = ['Paul', 'Ed', 'Kevin']
+
+    # Set seed to make this example reproducible
```
Comment (khwilson) [inline tests/test_suppression.py]: Would suggest creating a fixture:

```python
@pytest.fixture(scope="function")
def random() -> np.random.Generator:
    return np.random.default_rng(123456)
```

File: tests/test_suppression.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import numpy as np
 import pandas as pd
 import pytest
+import random
```
Comment (khwilson) [inline tests/test_suppression.py]: Don't need this import if you move to numpy

File: tests/test_suppression.py [hunk 6]
```diff
@@ -132,3 +133,26 @@ def test_single_row():
...
     assert (mask.values == expected_array).all()
+
+def test_nan_in_mean_sd_table():
+    total_length = 20
+    cuisine = ['Chinese', 'Korean', 'Italian']
+    city = ['Boston', 'Providence']
+    who = ['Paul', 'Ed', 'Kevin']
+
+    # Set seed to make this example reproducible
+    d = {"city": random.choices(city, k=total_length),
```
Comment (khwilson) [inline tests/test_suppression.py]: Is meant to deliberately create empty groupings or not? As it stands, I can't tell what the input to the test will be, which is somewhat critical when dealing with thresholds. :-)

**Review by khwilson — COMMENTED** [review]
Some small comments

File: tests/test_suppression.py [hunk 7]
```diff
@@ -132,3 +133,26 @@ def test_single_row():
...
+         "who": random.choices(who, k=total_length)
+         }
+    df = pd.DataFrame(d)
+
+    outcome_df = df.zen.mean_sd_table(
+        index=['cuisine', 'who'], columns='city', values='price', suppress=True, high=2
+    )
+    # Ensure that those that have '0' values have 'NA' values in sd
+    null_index = np.where(outcome_df[outcome_df.columns[outcome_df.columns.get_level_values(1) == 'n']]==0.0)
+    nan_index = np.where(outcome_df[outcome_df.columns[outcome_df.columns.get_level_values(1) == 'Mean (SD)']]== 'N/A')
```
Comment (digicosmos86) [inline tests/test_suppression.py]: You can also write `outcome_df.xs("n", level=1, axis=1) == 0.0` here

File: src/zentables/__init__.py (hunk 1 above)
Comment (edwardhuh) [reply src/zentables/__init__.py]: I thought this was already being achieved by removing `_do_suppression` from `__all__`.
If I remove `_do_suppression` from this original import, then there is no way for me to apply this function to `test_suppression.py`, because it never gets imported. 

I may have just completely misunderstood you, but I thought this was the move that was 'condoned' given the amount of testing we had to do.

File: src/zentables/zentables.py (hunk 2 above)
Comment (edwardhuh) [reply src/zentables/zentables.py]: Completely true. This was a bad call.

File: src/zentables/zentables.py (hunk 3 above)
Comment (edwardhuh) [reply src/zentables/zentables.py]: Done.

File: src/zentables/__init__.py (hunk 1 above)
Comment (digicosmos86) [reply src/zentables/__init__.py]: Ah, this was my bad. To test `_do_suppression` you need to do `from .zentables import _do_suppression` at the top of test scripts instead of `import zentables` which was at the beginning of my test script. Then you don't need to make `_do_suppression` accessible from the top level of the package.

**Comment by edwardhuh** [comment]
@khwilson I am realizing that I will need greater discussion on the testing-driven development :D

File: tests/test_suppression.py [hunk 8]
```diff
@@ -1,11 +1,15 @@
 import numpy as np
 import pandas as pd
 import pytest
-import random
 
 import zentables as zen
```
Comment (digicosmos86) [inline tests/test_suppression.py]: You can do `from .zentables import _do_suppression` here and avoid having `_do_suppression` as a top-level export, especially since `_do_suppression` begins with a `_` which indicates that it should never be exposed to users.

File: src/zentables/__init__.py [hunk 9]
```diff
@@ -13,6 +13,6 @@
     df.zen.pretty()
 """
 
-from .zentables import ZenTablesAccessor, set_options, _do_suppression
+from .zentables import ZenTablesAccessor, _do_suppression, set_options
```
Comment (digicosmos86) [inline src/zentables/__init__.py]: I would still remove `_do_suppression` here. See my other comments on making this work in test scripts

File: src/zentables/zentables.py [hunk 10]
```diff
@@ -614,7 +614,9 @@ def _internal_pivot_table(
             ).reset_index()
 
             if margins:
-                submargin = submargin.iloc[:-1,]
+                submargin = submargin.iloc[
```
Comment (digicosmos86) [inline src/zentables/zentables.py:617]: Should this be `submargin.iloc[:-1, :]`? Looks like R syntax to me. If I wrote this then my bad lol

**Review by digicosmos86 — COMMENTED** [review]
Really minor suggestions

**Comment by khwilson** [comment]
Hah @edwardhuh. The basic idea is to write your tests before your write your code. Some people are really rigorous about this, but I find it most useful when there's some sort of niggling technical issue --- a corner case, an unthought about class of cases --- where the explanation takes up a lot of words but an example is very short.

Then the example you just turn into a test . :-)
